cpp/src/stream_compaction/distinct.cu

```diff
@@ -37,18 +32,161 @@
+    auto const iter = d_map.find(idx, d_hasher, d_eqcomp);
+
+    // Only one index value of the duplicate rows could be inserted into the map.
+    // As such, looking up for all indices of duplicate rows always returns the same value.
+    auto const inserted_idx = iter->second.load(cuda::std::memory_order_relaxed);
+
+    // All duplicate rows will have concurrent access to the same output slot.
+    return d_output[inserted_idx];
+  }
+};
+
+}  // namespace
+
+rmm::device_uvector<size_type> get_distinct_indices(table_view const& input,
+                                                    duplicate_keep_option keep,
+                                                    null_equality nulls_equal,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  if (input.num_rows() == 0 or input.num_columns() == 0) {
+    return rmm::device_uvector<size_type>(0, stream, mr);
+  }
+
+  auto const preprocessed_input =
+    cudf::experimental::row::hash::preprocessed_table::create(input, stream);
+  auto const has_null = nullate::DYNAMIC{cudf::has_nested_nulls(input)};
+
+  auto map = hash_map_type{compute_hash_table_size(input.num_rows()),
+                           cuco::sentinel::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
+                           cuco::sentinel::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
+                           detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                           stream.value()};
+
+  auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
+  auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_null));
+
+  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
+  auto const key_equal = row_comp.equal_to(has_null, nulls_equal);
+
+  auto const kv_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
+  map.insert(kv_iter, kv_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+
+  // The output distinct indices.
+  auto output_indices = rmm::device_uvector<size_type>(map.get_size(), stream, mr);
+
+  // If we don't care about order, just gather indices of distinct keys taken from map.
+  if (keep == duplicate_keep_option::KEEP_ANY) {
+    map.retrieve_all(output_indices.begin(), thrust::make_discard_iterator(), stream.value());
+    return output_indices;
+  }
+
+  // Perform reduction on each group of rows compared equal and the results are store
+  // into this array. The reduction operator is:
+  // - If KEEP_FIRST: min.
+  // - If KEEP_LAST: max.
+  // - If KEEP_NONE: count number of appearances.
+  auto reduction_results = rmm::device_uvector<size_type>(input.num_rows(), stream);
+
+  auto const init_value = [keep] {
```

Comment:
`init_idx` or `init_index`

Reply:
Unfortunately `_value` here is not only for index meaning but also for counting (in case of `KEEP_NONE`) so `_index` is not suitable here.